### PR TITLE
changed precommit checks from shell to python

### PR DESCRIPTION
**Pull request by TeodoroFilippini**

### CODE REVIEW

File: precommit.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
+"""
```
Comment (mristin) [inline precommit.py]: Please keep consistent with pydocstyle (both `"""` on the same line, imperative mood)

File: precommit.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+Runs precommit checks on the repository.
+"""
+import pathlib
+import subprocess
+
+
+def main() -> None:
+    """"
+    Main routine
```
Comment (mristin) [inline precommit.py]: Ditto

File: precommit.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    """
+    root_dir = pathlib.Path(__file__).parent
+    source_dir = root_dir / "src"
+
+    retcode = subprocess.call(['which', 'elm-format'], stdout=subprocess.PIPE)
```
Comment (mristin) [inline precommit.py]: `stdout=subprocess.DEVNULL`, otherwise there's a memory leak since you never read from it.

File: precommit.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    if retcode != 0:
+        raise FileNotFoundError("elm-format (https://github.com/avh4/elm-format) not found.")
+    retcode = subprocess.call(['elm-format', '--elm-version=0.18', source_dir.as_posix(), '--yes'])
```
Comment (mristin) [inline precommit.py]: I'd just `check_call` here:

```
print("elm-format'ing...")
subprocess.check_call(['elm-format', '--elm-version=0.18', source_dir.as_posix(), '--yes'])
```

Also for the rest.

File: precommit.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    proc = subprocess.Popen(
+        cmd, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, universal_newlines=True, cwd=root_dir.as_posix())
+
+    out, _ = proc.communicate()
+    if proc.returncode != 0:
+        raise RuntimeError("elm-analyse failed. Did you install it :\n{}".format(out))
+    else:
+        print("Succeeded: elm-analyse")
+
+    retcode = subprocess.call(['which', 'elm-doc-test'], stdout=subprocess.PIPE)
```
Comment (mristin) [inline precommit.py]: I would group all `which` commands at the begining to fast-fail if any of the programs are missing.

File: precommit.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    if retcode != 0:
+        raise RuntimeError("Failed to elm-format: {}".format(source_dir))
+    else:
+        print("Succeeded: elm-format {}/...".format(source_dir.name))
+
+    retcode = subprocess.call(['which', 'elm-analyse'], stdout=subprocess.PIPE)
+    if retcode != 0:
+        raise FileNotFoundError("elm-analyse (https://stil4m.github.io/elm-analyse/) not found.")
+    cmd = ['elm-analyse']
+    proc = subprocess.Popen(
```
Comment (mristin) [inline precommit.py]: I would also just `check_call`, with `stdout=subprocess.DEVNULL`. Please see also for the rest of the commands.

**Comment by TeodoroFilippini** [comment]
@mristin sorry for the poor quality of the code. Does it look better now? :)

**Pull request merged**